client/mixnet/src/maybe_inf_delay.rs

```diff
@@ -0,0 +1,111 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use futures::{future::FusedFuture, FutureExt};
+use futures_timer::Delay;
+use std::{
+	future::Future,
+	pin::Pin,
+	task::{Context, Poll, Waker},
+	time::Duration,
+};
+
+enum Inner {
+	Infinite {
+		/// Waker from the most recent `poll` call. If `None`, either `poll` has not been called
+		/// yet, we returned `Poll::Ready` from the last call, or the waker is attached to `delay`.
+		waker: Option<Waker>,
+		delay: Option<Delay>,
+	},
+	Finite(Delay),
+}
+
+/// Like [`Delay`] but the duration can be infinite (in which case the future will never fire).
+/// Unlike [`Delay`], implements [`FusedFuture`], with [`is_terminated`](Self::is_terminated)
+/// returning `true` when the delay is infinite. As with [`Delay`], once [`poll`](Self::poll)
+/// returns [`Poll::Ready`], it will continue to do so until [`reset`](Self::reset) is called.
+pub struct MaybeInfDelay(Inner);
```

Comment:
Yeah true. This was originally needed for the mixnet libp2p behaviour, but that's gone now and the `reset` behaviour is not needed for `sc_mixnet::run`. I used it there because it's likely a bit more efficient than the alternative (I believe resetting a `Delay` is a bit more efficient than discarding and recreating), but I agree that it's not worth all this extra code.

Reply:
Removed.

Reply:
Hmm actually scratch that. `Fuse<Delay>` does not have quite the right behaviour. Unlike `MaybeInfDelay`, it will only return `Ready` once when the delay elapses; after that, it will return `Pending` again. This does not mesh well with the way these delays are reset. The mixnet crate sets flags when the delays need to change. If there are say two messages in a row with the same timestamp, it is possible that the mixnet crate will not set the delay-needs-to-change flag between them. This is fine with the current code as the delay will trigger again. This would _not_ work properly with `Fuse<Delay>`.